ports/atmel-samd/common-hal/pulseio/PulseOut.c

```diff
@@ -70,55 +69,78 @@ void pulse_finish(struct tc_module *const module) {
         return;
     }
     current_compare = (current_compare + pulse_buffer[pulse_index] * 3 / 4) & 0xffff;
-    tc_set_compare_value(MP_STATE_VM(pulseout_tc_instance), TC_COMPARE_CAPTURE_CHANNEL_0, current_compare);
+    Tc* tc = tc_insts[pulseout_tc_index];
+    tc->COUNT16.CC[0].reg = current_compare;
     if (pulse_index % 2 == 0) {
         turn_on(active_pincfg);
     }
 }
 
+void pulseout_interrupt_handler(uint8_t index) {
+    if (index != pulseout_tc_index) return;
+    Tc* tc = tc_insts[index];
+    if (!tc->COUNT16.INTFLAG.bit.MC0) return;
+
+    pulse_finish();
+
+    // Clear the interrupt bit.
+    tc->COUNT16.INTFLAG.reg = TC_INTFLAG_MC0;
+}
+
 void pulseout_reset() {
     refcount = 0;
-    MP_STATE_VM(pulseout_tc_instance) = NULL;
+    pulseout_tc_index = 0xff;
     active_pincfg = NULL;
 }
 
 void common_hal_pulseio_pulseout_construct(pulseio_pulseout_obj_t* self,
                                             const pulseio_pwmout_obj_t* carrier) {
     if (refcount == 0) {
         // Find a spare timer.
-        Tc *t = NULL;
-        Tc *tcs[TC_INST_NUM] = TC_INSTS;
-        for (uint8_t i = TC_INST_NUM; i > 0; i--) {
-            if (tcs[i - 1]->COUNT16.CTRLA.bit.ENABLE == 0) {
-                t = tcs[i - 1];
+        Tc *tc = NULL;
+        uint8_t index = TC_INST_NUM - 1;
+        for (; index >= 0; index--) {
```

Comment:
`index` is a `uint8_t`. So I don't see that this loop will ever finish.

Reply:
well, I see the break, but the condition in the for loop is always true.

Reply:
Good catch! I would have needed to test that it works when all TCs are already in use.